go/vt/topo/tablet_test.go

```diff
@@ -34,43 +34,65 @@ import (
 // GetTabletsByCell first tries to get all the tablets using List.
 // If the response is too large, we will get an error, and fall back to one tablet at a time.
 func TestServerGetTabletsByCell(t *testing.T) {
+	const cell = "zone1"
+	const keyspace = "keyspace"
+	const shard = "shard"
+
 	tests := []struct {
-		name      string
-		tablets   int
-		opt       *topo.GetTabletsByCellOptions
-		listError error
+		name               string
+		createShardTablets int
+		opt                *topo.GetTabletsByCellOptions
+		listError          error
+		keyspaceShards     map[string]string
 	}{
 		{
-			name:    "negative concurrency",
-			tablets: 1,
+			name:               "negative concurrency",
+			keyspaceShards:     map[string]string{keyspace: shard},
+			createShardTablets: 1,
 			// Ensure this doesn't panic.
 			opt: &topo.GetTabletsByCellOptions{Concurrency: -1},
 		},
 		{
-			name:    "single",
-			tablets: 1,
+			name:               "single",
+			keyspaceShards:     map[string]string{keyspace: shard},
+			createShardTablets: 1,
 			// Make sure the defaults apply as expected.
 			opt: nil,
 		},
 		{
-			name: "multiple",
+			name:           "multiple",
+			keyspaceShards: map[string]string{keyspace: shard},
 			// should work with more than 1 tablet
-			tablets: 32,
-			opt:     &topo.GetTabletsByCellOptions{Concurrency: 8},
+			createShardTablets: 32,
+			opt:                &topo.GetTabletsByCellOptions{Concurrency: 8},
 		},
 		{
-			name: "multiple with list error",
+			name:           "multiple with list error",
+			keyspaceShards: map[string]string{keyspace: shard},
 			// should work with more than 1 tablet when List returns an error
-			tablets:   32,
-			opt:       &topo.GetTabletsByCellOptions{Concurrency: 8},
-			listError: topo.NewError(topo.ResourceExhausted, ""),
+			createShardTablets: 32,
+			opt:                &topo.GetTabletsByCellOptions{Concurrency: 8},
+			listError:          topo.NewError(topo.ResourceExhausted, ""),
+		},
+		{
+			name: "filtered by keyspace and shard",
+			keyspaceShards: map[string]string{
+				keyspace:   shard,
+				"filtered": "-",
+			},
+			// should create 2 tablets in 2 different shards (4 total)
+			// but only a single shard is returned
```

Comment:
IMO it's worth making this more explicit in the test. For example:
```diff
diff --git a/go/vt/topo/tablet_test.go b/go/vt/topo/tablet_test.go
index c61e806579..5ac7d05d39 100644
--- a/go/vt/topo/tablet_test.go
+++ b/go/vt/topo/tablet_test.go
@@ -19,6 +19,7 @@ package topo_test
 import (
        "context"
        "errors"
+       "fmt"
        "testing"

        "github.com/stretchr/testify/assert"
@@ -145,10 +146,18 @@ func TestServerGetTabletsByCell(t *testing.T) {
                                require.Equal(t, tab.Tablet, tablets[i].Tablet)
                        }

+                       shards := make(map[string]struct{}, len(tt.keyspaceShards))
                        for _, tablet := range out {
+                               shards[fmt.Sprintf("%s/%s", tablet.Keyspace, tablet.Shard)] = struct{}{}
                                require.Equal(t, keyspace, tablet.Keyspace)
                                require.Equal(t, shard, tablet.Shard)
                        }
+                       if tt.opt != nil && tt.opt.KeyspaceShard != nil {
+                               t.Logf("Filtering shards. Total shards: %d, matching shards: %d", len(tt.keyspaceShards), len(shards))
+                               require.Len(t, shards, 1)
+                       } else {
+                               require.Equal(t, len(tt.keyspaceShards), len(shards))
+                       }
                })
        }
 }
```

Reply:
That's assuming, however, that we don't support passing in an empty keyspace with a non empty shard though.

Reply:
Either way, probably even better to have a test case member that explicitly contains the expected matches. Then that can be verified, not only the count but that we got the expected results. Right now we're not actually testing and confirming the results when filtering is done.

Reply:
Here's an example of what I mean above:
```diff
diff --git a/go/vt/topo/tablet_test.go b/go/vt/topo/tablet_test.go
index c61e806579..4afcb625b4 100644
--- a/go/vt/topo/tablet_test.go
+++ b/go/vt/topo/tablet_test.go
@@ -44,6 +44,7 @@ func TestServerGetTabletsByCell(t *testing.T) {
 		opt                *topo.GetTabletsByCellOptions
 		listError          error
 		keyspaceShards     map[string]string
+		expectedShards     map[string]string
 	}{
 		{
 			name:               "negative concurrency",
@@ -90,6 +91,7 @@ func TestServerGetTabletsByCell(t *testing.T) {
 					Shard:    shard,
 				},
 			},
+			expectedShards: map[string]string{keyspace: shard},
 		},
 	}
 
@@ -145,10 +147,17 @@ func TestServerGetTabletsByCell(t *testing.T) {
 				require.Equal(t, tab.Tablet, tablets[i].Tablet)
 			}
 
+			shards := make(map[string]string, len(tt.keyspaceShards))
 			for _, tablet := range out {
+				shards[tablet.Keyspace] = tablet.Shard
 				require.Equal(t, keyspace, tablet.Keyspace)
 				require.Equal(t, shard, tablet.Shard)
 			}
+
+			if tt.opt != nil && tt.opt.KeyspaceShard != nil {
+				require.NotEmpty(t, tt.expectedShards)
+				require.Equal(t, tt.expectedShards, shards)
+			}
 		})
 	}
 }

```